### PR TITLE
Make the [Experimental] warning as error by default

### PR DESCRIPTION
**Pull request by jcouv**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/ObsoleteAttributeHelpers.cs [hunk 1]
```diff
@@ -190,7 +190,7 @@ static DiagnosticInfo createObsoleteDiagnostic(Symbol symbol, BinderFlags locati
                     Debug.Assert(!data.IsError);
 
                     // Provide an explicit format for fully-qualified type names.
-                    return new CustomObsoleteDiagnosticInfo(MessageProvider.Instance, (int)ErrorCode.WRN_Experimental, data,
+                    return new CustomObsoleteDiagnosticInfo(MessageProvider.Instance, (int)ErrorCode.WRN_Experimental, effectiveSeverity: DiagnosticSeverity.Error, data,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/ObsoleteAttributeHelpers.cs:193]: @jaredpar Here's the core of the change for this design (being discussed in email). When producing WRN_Experimental, we're elevating the effective severity (before applying any of the command-line or editor config rules).

File: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs [hunk 2]
```diff
@@ -189,6 +192,13 @@ internal static class CSharpDiagnosticFilter
                     specifiedWarnAsErrorMinus = true;
                 }
             }
+            else if (errorCode == (int)ErrorCode.WRN_Experimental)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs:195]: @jaredpar @RikkiGibson  This approach should avoid any risk of impact on semantics (we delay the special handling until we're ready to report). 
We could make this logic conditional on `/warnaserror:experimental`, but that still only adds complexity with little value (doesn't avoid the special knowledge of this error code) so I prefer the unconditional approach.

File: src/Compilers/CSharp/Test/Emit2/Semantics/ExperimentalAttributeTests.cs [hunk 3]
```diff
@@ -2076,9 +2074,9 @@ void M(C c)
         if (inSource)
         {
             comp.VerifyDiagnostics(
-                // 0.cs(3,12): warning DiagID1: 'C' is for evaluation purposes only and is subject to change or removal in future updates.
+                // 0.cs(3,12): error DiagID1: 'C' is for evaluation purposes only and is subject to change or removal in future updates.
                 //     void M(C c)
-                Diagnostic("DiagID1", "C").WithArguments("C").WithLocation(3, 12)
+                Diagnostic("DiagID1", "C").WithArguments("C").WithLocation(3, 12).WithWarningAsError(true)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit2/Semantics/ExperimentalAttributeTests.cs:2079]: Do we have a test verifying that .editorconfig can suppress the diagnostic?

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 4]
```diff
@@ -1484,7 +1484,7 @@ internal enum ErrorCode
         ERR_FeatureNotAvailableInVersion7_1 = 8302,
         ERR_LanguageVersionCannotHaveLeadingZeroes = 8303,
         ERR_CompilerAndLanguageVersion = 8304,
-        WRN_Experimental = 8305,
+        WRN_WindowsExperimental = 8305,
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1487]: Is there an even more obscure name we can use for this? e.g. UWPExperimental? (I'm not sure if the usage is limited to there.)

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:1487]: We also use the "WindowsExperimental" name in `ObsoleteAttributeData.WindowsExperimental` and `ObsoleteAttributeKind.WindowsExperimental`, so I'd rather keep them all aligned.

File: src/Compilers/CSharp/Test/Emit2/Semantics/ExperimentalAttributeTests.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit2/Semantics/ExperimentalAttributeTests.cs:2079]: Added tests for editorconfig and nowarn. That caught a bug. Thanks!

File: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs [hunk 5]
```diff
@@ -214,7 +217,17 @@ internal static class CSharpDiagnosticFilter
 
             if (!isSpecified)
             {
-                report = isEnabledByDefault ? ReportDiagnostic.Default : ReportDiagnostic.Suppress;
+                if (errorCode == (int)ErrorCode.WRN_Experimental)
+                {
+                    // Special handling for [Experimental] warning (treat as error severity by default)
+                    severity = DiagnosticSeverity.Error;
+                    report = ReportDiagnostic.Error;
+                    isSpecified = true;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs:225]: Why do we need to set `isSpecified = true` here?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs]: You're right, it's not necessary/useful

File: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs [hunk 6]
```diff
@@ -214,7 +217,16 @@ internal static class CSharpDiagnosticFilter
 
             if (!isSpecified)
             {
-                report = isEnabledByDefault ? ReportDiagnostic.Default : ReportDiagnostic.Suppress;
+                if (errorCode == (int)ErrorCode.WRN_Experimental)
+                {
+                    // Special handling for [Experimental] warning (treat as error severity by default)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs:222]:
```suggestion
                    // Special handling for [Experimental] warning (treat as error severity by default)
                    Debug.Assert(isEnabledByDefault);
                    Debug.Assert(!specifiedWarnAsErrorMinus);
```

Both of those should hold correct?

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 7]
```diff
@@ -14701,6 +14701,139 @@ class C {}
...
+}
+
+[System.Diagnostics.CodeAnalysis.Experimental("DiagID1")]
+public class C
+{
+    public static void M() { }
+}
+""";
+
+            var comp = CreateCompilation(src, options: TestOptions.DebugExe.WithGeneralDiagnosticOption(ReportDiagnostic.Suppress));
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:14834]: >WithGeneralDiagnosticOption(ReportDiagnostic.Suppress) [](http://example.com/codeflow?start=76&length=54)

:memo: Although it cannot be specified from the command-line (`/nowarn` requires to specify some diagnostic IDs), it's possible to do a global suppression via APIs. Fixed this scenario

**Review by RikkiGibson — APPROVED** [review]

**Review by jaredpar — APPROVED** [review]

**Pull request merged**